### PR TITLE
[Raisin Bread] instruments and data

### PR DESCRIPTION
**Pull request by ridz1208**
Remake of https://github.com/aces/Loris/pull/3788 and https://github.com/aces/Loris/pull/3600

this PR does much more than it should but it is here to see if travis fails when the data is modified

SQL Changes: RBdata.sql
 - resolved discrepency between subprojects being `11,12,13,14` in some places and `1,2,3,4` in other places
 - made sure every instrument has a table in sql and metainformation. removed all entries in battery that have no instrument, removed any instrument that has no table, removed any table that has no instrument.
 - broken down to multiple files easier to modify when necessary in PRs

Instrument changes:
- added meta files for each example instrument
- updated instruments to new version when possible
- added sql create tables
- moved testing instruments to test folder

### CODE REVIEW

**Comment by johnsaigle** [comment]
Travis error: `ERROR 1054 (42S22) at line 1353: Unknown column 'PendingStaging' in 'field list'`

File: Dockerfile.test.db [hunk 1]
```diff
@@ -7,26 +7,31 @@ COPY SQL/0000-00-01-Permission.sql /0000-00-01-Permission.sql
...
     0000-00-04-Help.sql \
-    test_instrument.sql \
-    RBdata.sql \
+    aosi.sql \
+    bmi.sql \
     medical_history.sql \
-    radiology_review.sql > /docker-entrypoint-initdb.d/0000-compiled.sql 
+    mri_parameter_form.sql \
+    radiology_review.sql \
+    RBdata_meta_prerequesites.sql \
```
Comment (kongtiaowang) [inline Dockerfile.test.db]: RBdata_consent_info_history.sql  /

Comment (ridz1208) [reply Dockerfile.test.db]: Consent info history was removed in 20.0.0

I will update the docker

**Comment by ridz1208** [comment]
@kongtiaowang I made the changes, its still failing

**Comment by kongtiaowang** [comment]
@ridz1208 I need to make a PR for fixing the Travis and docker can't load config.xml and read the ajax file properly first, then we need to rebase this PR. I will fix this PR later.

**Comment by ridz1208** [comment]
TODO:

I need to make sure the visit_projec_subproject is properly populated

**Comment by kongtiaowang** [comment]
https://github.com/aces/Loris/pull/4145

**Comment by kongtiaowang** [comment]
test/unittests/VisitTest.php

**Comment by kongtiaowang** [comment]
@ridz1208 
https://github.com/aces/Loris/pull/4155/
Two files need to be modified.  
Database.class.inc  and test/unittests/VisitTest.php

File: docs/instruments/NDB_BVL_Instrument_TEMPLATE.class.inc [hunk 2]
```diff
@@ -1,89 +1,130 @@
...
+    {
+        if (preg_match("/<TEST_NAME>(_page[0-9]+)/", $this->page, $matches)) {
+            $this->_page($matches[1]);
         } else {
             $this->_main();
         }
-        //Defines the call back function for HTML Quickform to use when validating the form.
-        $this->form->addFormRule(array(&$this,'XINValidate'));
+        // Defines the call back function for HTML Quickform to use in validation
+        $this->form->addFormRule(array(&$this, 'XINValidate'));
```
Comment (paiva) [inline docs/instruments/NDB_BVL_Instrument_TEMPLATE.class.inc]: Shouldn't it be `([&$this, 'XINValidate'])` instead of `array(...)`?

Comment (johnsaigle) [reply docs/instruments/NDB_BVL_Instrument_TEMPLATE.class.inc]: No our standard for declaring arrays in PHp is using `array()` instead of `[]`.

**Comment by ridz1208** [comment]
TODO:

- [ ] Fix data in Project_rel table
- [ ] fill data in visit, visit_subproject_project tables
- [ ] update to version 20.1 + major changes

**Comment by kongtiaowang** [comment]
@ridz1208 
 Cannot truncate a table referenced in a foreign key constraint (`LorisTest`.`Config`, CONSTRAINT `fk_Config_1` FOREIGN KEY (`ConfigID`) REFERENCES `LorisTest`.`ConfigSettings` (`ID`))

**Comment by kongtiaowang** [comment]
test/RBfiles/RB_candidate.sql   =>  Gender to Sex

**Comment by kongtiaowang** [comment]
https://github.com/kongtiaowang/Loris/pull/208

File: SQL/0000-00-00-schema.sql [hunk 3]
```diff
@@ -359,7 +359,7 @@ CREATE TABLE `test_subgroups` (
 ) ENGINE=InnoDB DEFAULT CHARSET=utf8;
 
 
-INSERT INTO test_subgroups (Subgroup_name) VALUES ('Instruments');
+INSERT INTO test_subgroups (Subgroup_name) VALUES ('Instruments'),('Imaging');
```
Comment (ridz1208) [inline SQL/0000-00-00-schema.sql]: @driusan I added this to the schema becaue some of our "template" instruments assume it exists...

should I add it here or remove it from the instruments ?

**Comment by ridz1208** [comment]
@kongtiaowang I cleaned up the tables quite a bit and sourced all the new patches and clean-up patches.

I also changed the directory of theRB_files to put them under SQL

can you let me know why it's still failing now ?

**Comment by kongtiaowang** [comment]
ERROR 1146 (42S02) at line 81008: Table 'LorisTest.physiological_archive' doesn't exist

**Comment by kongtiaowang** [comment]
ERROR 1054 (42S22) at line 26892: Unknown column 'CenterID' in 'field list'

File: SQL/0000-00-00-schema.sql (hunk 3 above)
Comment (driusan) [reply SQL/0000-00-00-schema.sql]: I think it makes sense to include it here

**Comment by kongtiaowang** [comment]
RB_candidate.sql, switch CenterID to RegistrationCenterID

**Comment by ridz1208** [comment]
@kongtiaowang 

> ERROR 1054 (42S22) at line 27181: Unknown column 'CenterID' in 'field list'

Any way to know which table is generating that error ? I compared all the files in `RB_files` with the schema files and the columns CenterID match...

File: SQL/RB_files/RB_Config.sql [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+INSERT INTO `Config` (`ID`, `ConfigID`, `Value`) VALUES (28,32,'/data/');
+INSERT INTO `Config` (`ID`, `ConfigID`, `Value`) VALUES (29,33,'tools/logs/');
+INSERT INTO `Config` (`ID`, `ConfigID`, `Value`) VALUES (30,34,'/data/incoming/');
+INSERT INTO `Config` (`ID`, `ConfigID`, `Value`) VALUES (31,35,'/data/demo/bin/mri/');
+INSERT INTO `Config` (`ID`, `ConfigID`, `Value`) VALUES (32,36,'/data/incoming/');
+INSERT INTO `Config` (`ID`, `ConfigID`, `Value`) VALUES (33,37,'/PATH/TO/Genomic-Data/');
+INSERT INTO `Config` (`ID`, `ConfigID`, `Value`) VALUES (34,38,'/data/uploads/');
+INSERT INTO `Config` (`ID`, `ConfigID`, `Value`) VALUES (35,40,'main.css');
+INSERT INTO `Config` (`ID`, `ConfigID`, `Value`) VALUES (36,41,'25');
+INSERT INTO `Config` (`ID`, `ConfigID`, `Value`) VALUES (37,44,'abou-haider-dev.loris.ca');
```
Comment (driusan) [inline SQL/RB_files/RB_Config.sql]: I don't think this is right?

Comment (ridz1208) [reply SQL/RB_files/RB_Config.sql]: GRRRR

File: test/unittests/VisitTest.php [hunk 5]
```diff
@@ -53,37 +53,150 @@ protected function setUp()
...
-       $this->assertEquals($visit_name, $visit->getName(), "the name of the visit does not match");
+        $visit_name = "Visit 1";
+        $visit      = new Visit($visit_name);
+        $this->assertEquals($visit_name, $visit->getName(), "the name of the visit does not match");
     }
 
     function testAllVisit()
     {
         $visits = $this->_visitController->getAllVisits();
         $this->assertEquals($this->_listOfVisit, $visits, "the name of the visit does not match value in DB");
```
Comment (ridz1208) [inline test/unittests/VisitTest.php]: @driusan this assertequals seems to be comparing arrays in an ordered fashion which forced me to add an `order by` clause to the visitController for the array returned by the controller to match the one here.

I'm not sure thats normal behaviour...

Same goes for the one remaining error on the PR for function `testVisitsProjects`  the ordering seems to be the issue

File: docs/config/config.xml [hunk 6]
```diff
@@ -37,18 +37,18 @@
     <study>
         <PSCID>
             <!-- PSCID (Study Center ID) generation method possible options: sequential/random/user -->
-            <generation>sequential</generation>
+            <generation>random</generation>
             <structure>
                 <seq type="siteAbbrev"/>
-                <seq type="alphanumeric" minLength="4"/> <!-- Ex: AAA1-->
+                <seq type="numeric" length="3" min='100' max='999'/> <!-- Ex: AAA1-->
```
Comment (driusan) [inline docs/config/config.xml]: Why is this changed in the default config?

File: docs/config/config.xml [hunk 7]
```diff
@@ -61,27 +61,48 @@
...
+                <item value="V5">V5</item>
+                <item value="V6">V6</item>
             </labelSet>
         </visitLabel>
 
-         <!-- certification module -->
+        <!-- certification module -->
         <Certification>
-            <EnableCertification>0</EnableCertification>
+            <EnableCertification>1</EnableCertification>
```
Comment (driusan) [inline docs/config/config.xml]: I don't think this should be the default for LORIS, it makes it harder to set things up on a new install

**Comment by ridz1208** [comment]
@driusan makes sense for the config.xml. I'll try to find a solution where we can have a config for raisinbread alone and ideally not duplicate the default one.

as for the sourcing... not sure we need anything special. I'll maybe add a command to `cat` everything together into one big SQL

**Comment by ridz1208** [comment]
@driusan please re-review

**Comment by ridz1208** [comment]
@driusan I added some documentation. it could be improved later on if need be

**Comment by driusan** [comment]
@ridz1208 Where's the documentation? (The PR changes 183 files, so it's not obvious in the diff..)

**Comment by ridz1208** [comment]
@driusan its the only .md file

https://github.com/aces/Loris/pull/4120/files#diff-891d4b84b8f410eef353c238c8045ab6

File: raisinbread/README.md [hunk 8]
```diff
@@ -0,0 +1,56 @@
+*This directory contains all the files necessary to spawn an instance of the Raisin Bread demonstration dataset.*
```
Comment (driusan) [inline raisinbread/README.md]: can you format the paragraphs in this document with `fmt` to make it readable?

File: raisinbread/README.md [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+The RaisinBread data is stored in the form of SQL INSERT statements located in the `/raisinbread/RB_files/` directory and grouped by the database table they belong to. These statements rely on the pre-existence of the SQL tables and thus the data is heavily coupled with the default LORIS schema files located in the `/SQL/` directory. The RaisinBread dataset also include a few example instruments, these can be found in the `raisinbread/instruments/` directory along with their respective SQL schemas in `raisinbread/instruments/instrument_sql/` and their respective Meta SQL commands in `raisinbread/instruments/instrument_sql/Meta/`. 
+
+***Note:** The following instructions assume that the user has already installed a functional version of LORIS. It's important to note that the steps below will clear all data from the currently active database and replace it with the RaisinBread data.*
+
+***Assumption:** The commands below assume that the MySQL configuration files have exist and contain a default database, user, host and password. In the event that that is not the case, replace all `mysql` commands below by the necessary values `mysql -u user -p -h host database_name`*
```
Comment (driusan) [inline raisinbread/README.md]: "have exist"?

File: raisinbread/README.md [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+
+##### Sourcing SQL
+In order to be able to use the RaisinBread dataset, the LORIS SQL schema needs to be sourced, followed by the different instrument schemas and finally the actual RB data. The commands below assume that the current working directory is the main loris root directory.
```
Comment (driusan) [inline raisinbread/README.md]: s/loris/LORIS

File: raisinbread/README.md [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+
+1. Copy the `raisinbread/config.config.xml` file into `project/config.xml`
+2. Input the correct `<database>` information in the `project/config.xml` file 
+3. Change the values the the `Config` table of the SQL database to reflect the correct `host` and `base` values
+4. copy the `raisinbread/instruments/` instrument PHP and LINST files to the `projects/instruments/` directory
+
+### Modifying RB
+The RaisinBread database should be handled like any other project. The data should never be modified directly in the SQL files, the database should be sourced, altered and re-exported into SQL files.
+
+When possible, data modifications should be done directly from the LORIS front-end or using the loris tools. When modifications need to be done directly in SQL, it is important to only add clean data with proper realistic logical values where possible.
```
Comment (driusan) [inline raisinbread/README.md]: "loris" again.

Comment (driusan) [inline raisinbread/README.md]: I don't know what "proper realistic logical values" are.

File: raisinbread/README.md [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+
+
+### Exporting RB
+The RaisinBread data should be updated with every code change impacting the database. To do so, the user needs to source the database before the changes, modify the database and then export the data to be submitted back to LORIS in the same pull request as the code changes.
+
+##### Contributing back into LORIS
+In order to submit changes made to RaisinBread back to LORIS, an export tool was created in the `tools/exporters/` directory. The `DB_dump_table_data.php` PHP script executes an advanced `mysqldump` command that generates a single SQL file per SQL table in the `raisinbread/RB_files/` directory. Each file contains foreign keys disabling and enabling, table truncation to clear all existing data and table locking to avoid data corruption during sourcing. If the data has been modified between sourcing these files and exporting them, the modification will be reported as git uncommitted changes and thus they can be verified and submitted to the LORIS repo in the same pull request as the code.
```
Comment (driusan) [inline raisinbread/README.md]: What makes it "advanced"?

File: raisinbread/README.md (hunk 11 above)
Comment (ridz1208) [reply raisinbread/README.md]: noone does

File: raisinbread/README.md (hunk 12 above)
Comment (ridz1208) [reply raisinbread/README.md]: it has a bunch of parameters, replace with "complex" ?

**Comment by ridz1208** [comment]
@driusan done !!

File: raisinbread/README.md (hunk 12 above)
Comment (driusan) [reply raisinbread/README.md]: I don't think any extra adjective at all is required.

File: raisinbread/README.md [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+commands below by the necessary values `mysql -u user -p -h host
+database_name`*
+
+##### Sourcing SQL In order to be able to use the RaisinBread dataset, the
+LORIS SQL schema needs to be sourced, followed by the different instrument
+schemas and finally the actual RB data. The commands below assume that
+the current working directory is the main LORIS root directory.
+
+``` cat SQL/0000-00-00-schema.sql \
+    SQL/0000-00-01-Permission.sql \ SQL/0000-00-02-Menus.sql \
```
Comment (driusan) [inline raisinbread/README.md]: code block formatting is messed up.

File: raisinbread/README.md [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+    raisinbread/instruments/instrument_sql/mri_parameter_form.sql \
+    raisinbread/instruments/instrument_sql/radiology_review.sql \
+    test/test_instrument.sql \ raisinbread/RB_files/*.sql | mysql
+```
+
+##### Configuring In order to be able to load the LORIS front-end while
+using the RaisinBread dataset some configurations are necessary.
+
+1. Copy the `raisinbread/config.config.xml` file into `project/config.xml`
+2. Input the correct `<database>` information in the `project/config.xml`
```
Comment (driusan) [inline raisinbread/README.md]: list is missing line breaks

File: raisinbread/README.md [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+file 3. Change the values the the `Config` table of the SQL
+database to reflect the correct `host` and `base` values 4. copy the
+`raisinbread/instruments/` instrument PHP and LINST files to the
+`projects/instruments/` directory
+
+### Modifying RB The RaisinBread database should be handled like any other
```
Comment (driusan) [inline raisinbread/README.md]: This (and other headings) seems to be missing a line break now.

**Comment by ridz1208** [comment]
@driusan should look better now

**Comment by johnsaigle** [comment]
Also the `bids*` tables need to be dropped before the MRI tables. So move lines 116-119 above line 96.

**Comment by ridz1208** [comment]
@johnsaigle these can be fixed seperately, it has nothing to do with raisinbread its the default schemas

**Comment by ridz1208** [comment]
@johnsaigle I made the changes separately to avoid more files changed 
https://github.com/aces/Loris/pull/4459/files

**Comment by christinerogers** [comment]
Tagging "Document at Release" because moving bmi.linst and other files from docs/instruments also requires updating the Loris Setup Guide (wiki/markdown) for Behavioural instruments.
